nix/docker/k8s/cardano-nodes.yaml

```diff
@@ -0,0 +1,190 @@
+#
+# NAMESPACE=cardano
+# kubectl create namespace $NAMESPACE
+#
+
+# GCE Storage Class
+---
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  name: cardano-standard-rwo
+provisioner: kubernetes.io/gce-pd
+parameters:
+  type: pd-standard
+  fstype: ext4
+  replication-type: none
+
+# Relay ###############################################################################################################
+
+# Relay Volume Claim
+---
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: relay-data
+  labels:
+    app: cardano
+    pvc: relay-data
+spec:
+  storageClassName: cardano-standard-rwo
+  accessModes: [ ReadWriteOnce ]
+  resources:
+    requests:
+      storage: 10G
+
+# Relay Pod
+---
+apiVersion: v1
+kind: Pod
+metadata:
+  name: relay
+  labels:
+    app: cardano
+    pod: relay
+spec:
+  containers:
+  - name: relay
+    image: inputoutput/cardano-node:dev
+    imagePullPolicy: Always
+    env:
+    - name: CARDANO_PUBLIC_IP
+      valueFrom:
+        configMapKeyRef:
+          name: relaycfg
+          key:  publicIP
+    - name: CARDANO_TOPOLOGY
+      valueFrom:
+        configMapKeyRef:
+          name: relaycfg
+          key:  topology
+    - name: CARDANO_CUSTOM_PEERS
+      valueFrom:
+        configMapKeyRef:
+          name: relaycfg
+          key:  customPeers
+    - name: CARDANO_UPDATE_TOPOLOGY
+      value: "true"
+    args: [ "run" ]
+    ports:
+      - containerPort: 3001
+    volumeMounts:
+    - name: datavol
+      mountPath: /opt/cardano/data
+  nodeSelector:
+    nodeType: relay
+  volumes:
+  - name: datavol
+    persistentVolumeClaim:
+      claimName: relay-data
+
+# Relay TCP Service
+---
+apiVersion: v1
+kind: Service
+metadata:
+  name: relay-np
+  labels:
+    app: cardano
+spec:
+  type: NodePort
+  selector:
+    app: cardano
+    pod: relay
+  ports:
+  - name: tcp
+    port: 3001
+    nodePort: 30010
```

Comment:
I'm having trouble finding documentation on the ports and what needs to be exposed and how stateful the connections need to be. Do you have a suggestion/chunk of code I could look at? Trying to determine if we absolutely need NodePort which limits scaling w/i a node, or if there's a way we could expose these via a load balancer + ingress.

Reply:
Currently, there is a (central) topology updating service, which every node has to call exactly once per hour. A node is accepted after three successful calls with the same IP. Therefore, this IP has to be stable or be targeted by some stable DNS name.

This will change in the near future with Alonzo when Cardano nodes will be able to find their peers on their own (i.e. without the help of an external topology update service).

Reply:
Much of what you see here can probably be improved quite considerably by using a StatefulSet instead of individual Pods. This is what I'm currently looking at over at [nessus-cardano](https://github.com/tdiesler/nessus-cardano).

As for now, I'm not sure whether IOHK appreciates this work let alone plans to merge it upstream eventually. Your comments are the first piece of technical feedback that I got - and thanks a million for that.

Yes, please come over to  nessus-cardano to provide more feedback and help with improved versions of this. That project is not meant to be a fork of the iohk image, but rather an incubating space for future PRs like this.

The nessusio/cardano-node images are [here](https://hub.docker.com/r/nessusio/cardano-node). They have seen quite a bit of adoption already and hopefully can see a lot less when this tech becomes available here.